File: work/lib/t_strategy/portfolio/risk.py
"""
Risk Management Models — 风险管理模型
========================================
来源: 凯利, VaR, 波动率目标, 固定比例止损
"""
from __future__ import annotations

import numpy as np
from typing import Optional, Tuple


def volatility_target_position(
    current_vol: float,
    target_vol: float = 0.20,
    max_leverage: float = 1.0,
) -> float:
    """波动率目标仓位 (Volatility Targeting)。"""
    if current_vol <= 0:
        return 0
    return min(max_leverage, target_vol / current_vol)


def fixed_stop_loss(entry_price: float, current_price: float,
                    stop_loss_pct: float = 0.07, direction: str = "long") -> Tuple[bool, float]:
    """固定比例止损。

    Returns:
        (是否触发止损, 触发价格)
    """
    if direction == "long":
        stop_price = entry_price * (1 - stop_loss_pct)
        triggered = current_price <= stop_price
    else:
        stop_price = entry_price * (1 + stop_loss_pct)
        triggered = current_price >= stop_price
    return triggered, stop_price


def trailing_stop(current_price: float, peak_price: float,
                  trail_pct: float = 0.08) -> Tuple[bool, float]:
    """移动止损。

    Returns:
        (是否触发, 止损价)
    """
    stop_price = peak_price * (1 - trail_pct)
    triggered = current_price <= stop_price
    return triggered, stop_price


def position_sizing(
    account_value: float,
    risk_per_trade: float = 0.01,
    entry_price: float = 0,
    stop_price: float = 0,
) -> int:
    """基于固定风险比例的仓位计算。

    Args:
        account_value: 账户总资产
        risk_per_trade: 单笔可接受亏损比例 (默认1%)
        entry_price: 买入价
        stop_price: 止损价

    Returns:
        建议买入股数
    """
    if entry_price <= 0 or stop_price <= 0 or entry_price <= stop_price:
        return 0
    risk_per_share = entry_price - stop_price
    max_loss = account_value * risk_per_trade
    shares = int(max_loss / risk_per_share)
    return max(shares, 0)


def value_at_risk(returns: np.ndarray, confidence: float = 0.95) -> float:
    """VaR — 在险价值 (参数法)。"""
    mu = np.mean(returns)
    sigma = np.std(returns)
    from scipy.stats import norm
    return float(mu - sigma * norm.ppf(confidence))


def max_drawdown(prices: np.ndarray) -> Tuple[float, int, int]:
    """最大回撤。

    Returns:
        (最大回撤比例, 峰值索引, 谷值索引)
    """
    peak = np.maximum.accumulate(prices)
    drawdown = (prices - peak) / peak
    max_dd = np.min(drawdown)
    valley_idx = np.argmin(drawdown)
    peak_idx = np.argmax(prices[:valley_idx + 1])
    return float(max_dd), int(peak_idx), int(valley_idx)


def sharpe_ratio(returns: np.ndarray, rf: float = 0.02) -> float:
    """夏普比率。"""
    excess = returns - rf / 252
    return float(np.mean(excess) / np.std(excess) * np.sqrt(252))


def sortino_ratio(returns: np.ndarray, rf: float = 0.02) -> float:
    """索蒂诺比率 (只考虑下行波动)。"""
    excess = returns - rf / 252
    downside = np.std(excess[excess < 0])
    if downside == 0:
        return 0
    return float(np.mean(excess) / downside * np.sqrt(252))


def calmar_ratio(returns: np.ndarray, prices: np.ndarray) -> float:
    """卡玛比率 (年化收益 / 最大回撤)。"""
    ann_return = (prices[-1] / prices[0]) ** (252 / len(prices)) - 1
    dd, _, _ = max_drawdown(prices)
    if dd == 0:
        return 0
    return float(ann_return / abs(dd))


print("  ✔ portfolio/risk")

File: packages/t-strategy/src/portfolio/risk.ts
/** Risk management utilities */

export function volatilityTargetPosition(currentVol: number, targetVol = 0.2, maxLeverage = 1) {
  if (currentVol <= 0) return 0
  return Math.min(maxLeverage, targetVol / currentVol)
}

export function fixedStopLoss(
  entryPrice: number, currentPrice: number, stopLossPct = 0.07, direction: 'long' | 'short' = 'long',
) {
  if (direction === 'long') {
    const stopPrice = entryPrice * (1 - stopLossPct)
    return { triggered: currentPrice <= stopPrice, stopPrice }
  }
  const stopPrice = entryPrice * (1 + stopLossPct)
  return { triggered: currentPrice >= stopPrice, stopPrice }
}

export function trailingStop(currentPrice: number, peakPrice: number, trailPct = 0.08) {
  const stopPrice = peakPrice * (1 - trailPct)
  return { triggered: currentPrice <= stopPrice, stopPrice }
}

export function positionSizing(
  accountValue: number, riskPerTrade = 0.01, entryPrice = 0, stopPrice = 0,
) {
  if (entryPrice <= 0 || stopPrice <= 0 || entryPrice <= stopPrice) return 0
  const riskPerShare = entryPrice - stopPrice
  return Math.max(0, Math.floor((accountValue * riskPerTrade) / riskPerShare))
}

export function valueAtRisk(returns: number[], confidence = 0.95) {
  if (!returns.length) return 0
  const mu = returns.reduce((a, b) => a + b, 0) / returns.length
  const variance = returns.reduce((a, r) => a + (r - mu) ** 2, 0) / returns.length
  const sigma = Math.sqrt(variance)
  // z ≈ 1.645 for 95%
  const z = confidence >= 0.99 ? 2.326 : confidence >= 0.95 ? 1.645 : 1.282
  return mu - sigma * z
}

export function maxDrawdown(prices: number[]) {
  let peak = prices[0] ?? 0
  let maxDd = 0
  let peakIdx = 0
  let valleyIdx = 0
  for (let i = 0; i < prices.length; i++) {
    if (prices[i] > peak) { peak = prices[i]; peakIdx = i }
    const dd = peak > 0 ? (prices[i] - peak) / peak : 0
    if (dd < maxDd) { maxDd = dd; valleyIdx = i }
  }
  return { maxDrawdown: maxDd, peakIdx, valleyIdx }
}

export function sharpeRatio(returns: number[], rf = 0.02) {
  if (returns.length < 2) return 0
  const dailyRf = rf / 252
  const excess = returns.map(r => r - dailyRf)
  const mean = excess.reduce((a, b) => a + b, 0) / excess.length
  const std = Math.sqrt(excess.reduce((a, r) => a + (r - mean) ** 2, 0) / (excess.length - 1))
  return std === 0 ? 0 : (mean / std) * Math.sqrt(252)
}

export function sortinoRatio(returns: number[], rf = 0.02) {
  const dailyRf = rf / 252
  const excess = returns.map(r => r - dailyRf)
  const mean = excess.reduce((a, b) => a + b, 0) / excess.length
  const downside = excess.filter(r => r < 0)
  if (!downside.length) return 0
  const std = Math.sqrt(downside.reduce((a, r) => a + r ** 2, 0) / downside.length)
  return std === 0 ? 0 : (mean / std) * Math.sqrt(252)
}

export function calmarRatio(returns: number[], prices: number[]) {
  if (prices.length < 2) return 0
  const annReturn = (prices[prices.length - 1] / prices[0]) ** (252 / prices.length) - 1
  const { maxDrawdown: dd } = maxDrawdown(prices)
  return dd === 0 ? 0 : annReturn / Math.abs(dd)
}
